### ISSUE TITLE
Build failure: diffutils-3.10 armv6l-linux linux on qemu

### ISSUE THREAD

**Issue by Ramblurr** [NixOS/nixpkgs#302050]
### Steps To Reproduce

1. `nix build github:NixOS/nixpkgs/3541c7628de9414f999ced044c95a284584f4ec1#legacyPackages.armv6l-linux.diffutils`

### Build log

[f2xh157azy08gwdzz5c46l0jqhqvvv0d-diffutils-3.10.drv.txt](https://github.com/NixOS/nixpkgs/files/14892947/f2xh157azy08gwdzz5c46l0jqhqvvv0d-diffutils-3.10.drv.txt)

```

============================================================================
Testsuite summary for GNU diffutils 3.10
============================================================================
# TOTAL: 215
# PASS:  175
# SKIP:  37
# XFAIL: 0
# FAIL:  3
# XPASS: 0
# ERROR: 0
============================================================================
See gnulib-tests/test-suite.log
Please report to bug-diffutils@gnu.org
```

### Additional context

Possibly related?

* https://github.com/NixOS/nixpkgs/issues/300550
* https://github.com/NixOS/nixpkgs/pull/300571

### Notify maintainers

@dasJ
@alyssais

### Metadata

Please run `nix-shell -p nix-info --run "nix-info -m"` and paste the result.

```console
[user@system:~]$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"x86_64-linux"`
 - host os: `Linux 6.7.11, NixOS, 24.05 (Uakari), 24.05.20240329.d8fe5e6`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.18.2`
 - nixpkgs: `/nix/store/cb1gs888vfqxawvc65q1dk6jzbayh3wz-source`
```

---

Add a :+1: [reaction] to [issues you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[issues you find important]: https://github.com/NixOS/nixpkgs/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc

**Comment by Ramblurr** [comment]
@theotheroracle does it make sense to apply your fix to the armv6l platform too?

**Comment by alyssais** [comment]
Can't we fix the emulators?  I don't see how it's feasible for Nixpkgs to try to support arbitrary process-level emulators that do who knows how good a job of emulating the platform they're pretending to be.

**Comment by Ramblurr** [comment]
The PR for risc mentioned it was a qemu issue, but didn't have any links to upstream. Ensuring the bug is reported upstream would probably be a good thing to do.

That said the rationale for disabling the tests on risc apply to armv6l imo. Only masochists are building Nixpkgs on their raspberry pi 1s, it takes days or even weeks. The emulator is the sane way to do it. (Especially because armv6l isn't in the cache).

**Comment by alyssais** [comment]
> That said the rationale for disabling the tests on risc apply to armv6l imo.

There's an important difference.  For RISC-V, we expect that over time fewer people will be building using emulators as hardware gets more powerful, so disabling the tests to accommodate the emulator is a temporary hack.  For armv6l, I don't think that's the case, so it wouldn't be something we could get rid of later.

**Comment by theotheroracle** [comment]
> @theotheroracle does it make sense to apply your fix to the armv6l platform too?

i dont see why not

**Comment by theotheroracle** [comment]
> The PR for risc mentioned it was a qemu issue, but didn't have any links to upstream. Ensuring the bug is reported upstream would probably be a good thing to do.

yes true, but i dont really have the expertise personally to report it, as i dont fully understand the issue

**Comment by Ramblurr** [comment]
> > That said the rationale for disabling the tests on risc apply to armv6l imo.
> 
> There's an important difference. For RISC-V, we expect that over time fewer people will be building using emulators as hardware gets more powerful, so disabling the tests to accommodate the emulator is a temporary hack. For armv6l, I don't think that's the case, so it wouldn't be something we could get rid of later.

I think the opposite is the case, no? armv6l is getting more and more rare, it's very slow, so compiling a whole distro is a big ask. I expect most building for armv6l to be compiled with qemu.

I would love to report it upstream to qemu but I have no information to give them except that someone said it was a qemu bug. Without an evidence it's difficult to make an issue.

**Comment by alyssais** [comment]
> I think the opposite is the case, no? armv6l is getting more and more rare, it's very slow, so compiling a whole distro is a big ask. I expect most building for armv6l to be compiled with qemu.

That's exactly my point.  With RISC-V, there will be a point in the future where we can remove the hack, because people won't be using qemu for builds much.  For armv6l, people will probably be using QEMU forever, so the argument that it's a time-limited hack doesn't apply, unless we want to stop supporting armv6l entirely (which I don't think would be desirable).

**Comment by Ramblurr** [comment]
Ah yes then we agree on that. Sorry for the misunderstanding.

**Comment by theotheroracle** [comment]
when the issue is resolved with qemu the patch could be resolved no ?

**Comment by Ramblurr** [comment]
Yes if:

1. It is indeed a qemu bug
2. and qemu is informed about the bug
3. qemu cares about the bug and fixes it

So far all we have is @AlexandreTunstall's  [guess that it is a qemu bug](https://github.com/NixOS/nixpkgs/issues/300550#issuecomment-2032136788), but not enough info to [upstream it to qemu](https://www.qemu.org/contribute/report-a-bug/).

A good first step would be to try to reproduce the bug without nix tooling so that qemu devs can grok it.

**Comment by MostAwesomeDude** [comment]
I just ran into this (or #300550, whichever) targeting `armv7l-linux` from a `x86_64-linux` host. Hacking an overlay to not `doCheck` is sufficient to unbreak the build at that point:

        (final: prev: {
          diffutils = prev.diffutils.overrideAttrs (_: {
            doCheck = !prev.stdenv.buildPlatform.isArmv7;
          });
        })

Given that all three failures are related to C-style stack handling and stack overflow, that QEMU emulates that stack, and that this is a multi-architectural issue, it seems rather obvious that this is a bug in QEMU's arch-independent stack code. The main question is whether this is a regression, and I don't have the patience to `bisect` nixpkgs tonight.

**Comment by Ramblurr** [comment]
Any efforts to investigate if qemu is aware of this or not would be most welcome.
